### PR TITLE
feat: implement `-o` flags for disabling optimizations

### PR DESCRIPTION
**Pull request by kayagokalp**
## Description
This PR implements `-o` flags so that we can disable optimizations. This will be useful as we are working towards debugger support and optimizations pollute source map generated. 

Provide option to the e2e test binary to specify build profile. For tests that have their ABI tested, or rely on their compiled
hashes (for deployment), since that changes with build profile, the tests are marked as unsupported for debug profile.
(better to test with and have expected results for release than the debug profile).

Add testing release profile in CI since the default is now debug. Two tests in `should_fail` that are expected to fail due to
overflows are disabled because there's a bug in our IR gen. A new Issue will be opened for this.

### CODE REVIEW

**Comment by kayagokalp** [comment]
After discussing it with @vaivaswatha, this is blocked as some optimizations are currently needed for producing correct output, so we cannot directly chug them depending on the o flag, we will have to be more careful about that.

**Comment by vaivaswatha** [comment]
@kayagokalp I've updated the PR to disable as many optimizations as possible on `-O0` (which is the default path). Can you check if the location information has improved for you with this?

The testsuite mostly passes. The ones failing are related to changed ABIs or checksums (for deployment ones). 

This brings us to the task of supporting two paths in our testsuite, i.e., testing both `-O1` and `-O1`, which I believe should be part of this PR. We can get to it after you confirm that disabling some of these optimizations indeed did help you.

**Comment by kayagokalp** [comment]
Thanks for the help @vaivaswatha I will be reporting before and after changes here today 🙌

**Comment by vaivaswatha** [comment]
> Thanks for the help @vaivaswatha I will be reporting before and after changes here today 🙌

@kayagokalp does having fewer optimizations improve the quality of source location information?

File: test/src/e2e_vm_tests/test_programs/should_pass/language/references/referencing_local_vars_and_values/test.toml [hunk 1]
```diff
@@ -1,4 +1,4 @@
 category = "run"
 expected_result = { action = "return", value = 42 }
 validate_abi = true
-expected_warnings = 9 # TODO-IG: Should be zero. DCA problem? Investigate why there are "This method is never called." errors for some of the types.
+expected_warnings = 90 # TODO-IG: Should be zero. DCA problem? Investigate why there are "This method is never called." errors for some of the types.
```
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply test/src/e2e_vm_tests/test_programs/should_pass/language/references/referencing_local_vars_and_values/test.toml:4]: This is a mistake. Thank you for finding it. I'll revert this back.

File: test/src/e2e_vm_tests/test_programs/should_fail/arith_overflow/u8_add_overflow/test.toml [hunk 2]
```diff
@@ -1,3 +1,3 @@
-category = "run"
+category = "disabled"
```
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply test/src/e2e_vm_tests/test_programs/should_fail/arith_overflow/u8_add_overflow/test.toml:1]: This is tracked by https://github.com/FuelLabs/sway/issues/5449

File: forc-pkg/src/manifest.rs [hunk 3]
```diff
@@ -233,6 +233,7 @@ pub struct BuildProfile {
     #[serde(default)]
     pub error_on_warnings: bool,
     pub reverse_results: bool,
+    pub optimization_level: usize,
```
[the remark this answers is not in the document]
Comment (vaivaswatha) [reply forc-pkg/src/manifest.rs:236]: Perhaps. @kayagokalp do you have any comment? (asking since you made this part of the change).

**Review by sdankel — DISMISSED** [review]
Looks good! Thank you!!

**Comment by vaivaswatha** [comment]
@tritao if this is ok, except for  #5486 related changes, we can merge it and update that PR. Or we can wait till that is merged.

**Review by sdankel — APPROVED** [review]

**Pull request merged**